hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java

```diff
@@ -422,9 +422,21 @@ private void submitConnectToClusterTask() {
                             logger.warning("No connection to cluster: " + clusterId);
                         }
 
-                        submitConnectToClusterTask();
+                        submitConnectToClusterTask(initialConnection);
                     }
                 }
+                if (initialConnection && !activeConnections.isEmpty()) {
```

Comment:
I have been thinking about this part and I believe there is a problem here. When I enter this if block, it means the client is connected to the cluster and since this is the initial connection, the client state is set to `INITIALIZED_ON_CLUSTER`.

However, if there was a state to be sent, it is not sent yet. So, we will allow invocations before sending the client's state. Imagine a long pause here, before sending the state to the cluster. Then, even the `testAsyncStartTrueShouldNotBlock_whenThereIsClientStateToSend` might fail

Reply:
Right. What about changing the client connection logic so that: 

We don't make the client INITIALIZED_ON_CLUSTER on first connection if the client is asyncStart. The async state sender code will make the client INITIALIZED_ON_CLUSTER

Reply:
Yes, that might work.